movement/watch_faces/complication/totp_face_lfs.c

```diff
@@ -0,0 +1,231 @@
+#include <stdlib.h>
+#include <string.h>
+#include <math.h>
+
+#include "TOTP.h"
+#include "base32.h"
+
+#include "watch.h"
+#include "watch_utility.h"
+#include "filesystem.h"
+
+#include "totp_face_lfs.h"
+
+/* Reads from a file totp_uris.txt where each line is what's in a QR code:
+ * e.g.
+ *   otpauth://totp/Example:alice@google.com?secret=JBSWY3DPEHPK3PXP&issuer=Example
+ *   otpauth://totp/ACME%20Co:john.doe@email.com?secret=HXDMVJECJJWSRB3HWIZR4IFUGFTMXBOZ&issuer=ACME%20Co&algorithm=SHA1&digits=6&period=30
+ * This is also the same as what Aegis exports in plain-text format.
+ *
+ * Minimal sanitisation of input, however.
+ *
+ * At the moment, to get the records onto the filesystem, start a serial connection and do:
+ *   echo otpauth://totp/Example:alice@google.com?secret=JBSWY3DPEHPK3PXP&issuer=Example > totp_uris.txt
+ *   echo otpauth://totp/ACME%20Co:john.doe@email.com?secret=HXDMVJECJJWSRB3HWIZR4IFUGFTMXBOZ&issuer=ACME%20Co&algorithm=SHA1&digits=6&period=30 >> totp_uris.txt
+ * (note the double >> in the second one)
+ *
+ * You may want to customise the characters that appear to identify the 2FA code. These are just the first two characters of the issuer,
+ * and it's fine to modify the URI.
+ */
+
+
+#define MAX_TOTP_RECORDS 20
+#define TOTP_FILE "totp_uris.txt"
+
+const char* TOTP_URI_START = "otpauth://totp/";
+
+struct totp_record {
+    uint8_t secret[48];
+    size_t secret_size;
+    char label[2];
+    uint32_t period;
+};
+
+static struct totp_record totp_records[MAX_TOTP_RECORDS];
+static int num_totp_records = 0;
+
+static void init_totp_record(struct totp_record *totp_record) {
+    totp_record->secret_size = 0;
+    totp_record->label[0] = 'A';
+    totp_record->label[1] = 'A';
+    totp_record->period = 30;
+}
+
+static bool totp_face_lfs_read_param(struct totp_record *totp_record, char *param, char *value) {
+    if (!strcmp(param, "issuer")) {
+        if (value[0] == '\0' || value[1] == '\0') {
+            printf("TOTP issuer must be >= 2 chars, got '%s'\n", value);
+            return false;
+        }
+        totp_record->label[0] = value[0];
+        totp_record->label[1] = value[1];
+    } else if (!strcmp(param, "secret")) {
+        totp_record->secret_size = base32_decode((unsigned char *)value, totp_record->secret);
```

Comment:
I feel like checking BASE32_LEN is the responsible thing to do here... (i.e. make sure it's <= 48, now that the max secret size is fixed)

Reply:
This is also a good call, I can make this change.